tools/serverpod_cli/lib/src/language_server/language_server.dart

```diff
@@ -1,20 +1,93 @@
+import 'dart:async';
 import 'dart:io';
-import 'package:json_rpc_2/json_rpc_2.dart';
-import 'package:serverpod_cli/src/language_server/wireformat.dart';
+import 'package:lsp_server/lsp_server.dart';
+import 'package:serverpod_cli/src/language_server/diagnostics_source.dart';
+
+import '../../analyzer.dart';
+import '../generator/code_generation_collector.dart';
 
 Future<void> runLanguageServer() async {
-  var peer = Peer(lspChannel(stdin, stdout));
-  peer
-    ..registerMethod('initialize', (params) async {
-      return ({
-        'capabilities': {
-          'textDocumentSync': 1,
-        }
-      });
-    })
-    ..registerMethod('textDocument/didChange', (Parameters params) async {
-      // todo implement
+  var connection = Connection(stdin, stdout);
+
+  connection.onInitialize((params) async {
+    return InitializeResult(
+      capabilities: ServerCapabilities(
+        textDocumentSync: const Either2.t1(TextDocumentSyncKind.Full),
+      ),
+    );
+  });
+
+  connection.onDidOpenTextDocument((params) async {
+    var diagnostics = _validateTextDocument(
+      params.textDocument.text,
+      params.textDocument.uri.toString(),
+    );
+
+    connection.sendDiagnostics(
+      PublishDiagnosticsParams(
+        diagnostics: diagnostics,
+        uri: params.textDocument.uri,
+      ),
+    );
+  });
+
+  connection.onDidChangeTextDocument((params) async {
+    var contentChanges = params.contentChanges.map((content) {
+      return content.map(
+        (document) => TextDocumentContentChangeEvent2(text: document.text),
+        (document) => document,
+      );
     });
 
-  await peer.listen();
+    var diagnostics = _validateTextDocument(
+      contentChanges.last.text,
+      params.textDocument.uri.toString(),
+    );
+
+    connection.sendDiagnostics(
+      PublishDiagnosticsParams(
+        diagnostics: diagnostics,
+        uri: params.textDocument.uri,
+      ),
+    );
+  });
+
+  await connection.listen();
+}
+
+List<Diagnostic> _validateTextDocument(String yaml, String sourcePath) {
```

Comment:
Naming should reflect that it validates a SerializableEntity and not a generic text document?